pre_commit/editor.py

```diff
@@ -0,0 +1,150 @@
+import psutil
+import argparse
+import shlex
+import contextlib
+import concurrent.futures
+import subprocess
+import os
+from pathlib import Path
+from typing import Any, Callable, List, NoReturn, TypeVar
+
+from pre_commit.metrics import monitor
+from pre_commit import output, git
+
+T = TypeVar('T')
+
+# Keeping the file name the same as git's makes it more likely that editors will set the file type
+# correctly when opening it.
+COMMIT_MESSAGE_DRAFT_PATH = Path('.git/pre-commit/COMMIT_EDITMSG')
+COMMIT_MESSAGE_HEADER = """
+# Please enter the commit message for your changes. Lines starting
+# with '#' will be ignored, and an empty message aborts the commit.
+#
+"""
+
+
+def should_run_concurrently(hook_stage: str) -> bool:
+    return hook_stage == 'commit' and _should_open_editor() and _is_editor_script_configured()
+
+def run_concurrently(fun: Callable[..., T], *args: Any) -> T:
+    # Allow user to enter commit message concurrently with running pre-commit hooks to lower
+    # wait times.
+    output.write_line('Waiting for your editor (pre-commit hooks running in background)...')
+
+    # Run git commands before starting hooks to avoid race conditions and git lock errors.
+    commit_message_template = _get_commit_message_template()
+
+    with contextlib.ExitStack() as paused_stdout_stack:
+        paused_stdout_stack.enter_context(output.paused_stdout())
+        def launch_editor() -> None:
+            _edit_commit_message(commit_message_template)
+            paused_stdout_stack.close()  # Resume terminal output as soon as the editor closes
+        with concurrent.futures.ThreadPoolExecutor(max_workers=2) as ex:
+            ex.submit(launch_editor)
+            retval_future = ex.submit(fun, *args)
+            return retval_future.result()
+    raise RuntimeError('unreachable')
+
+
+class ParseFailed(BaseException):
+    pass
+
+class NoExitParser(argparse.ArgumentParser):
+    # exit_on_error option only exists in 3.9+, so we have to do this ourselves.
+    def error(self, _: str) -> NoReturn:
+        raise ParseFailed()
+
+
+def _should_open_editor() -> bool:
+    git_invocation = psutil.Process().parent().cmdline()
+    if git_invocation[:2] != ['git', 'commit']:
+        # Some other command is being run; let's be conservative.
+        return False
+    try:
+        # Teach the parser about all the allowable arguments to git commit and have it fail if any
+        # others are present.
```

Comment:
Not a blocker: This seems like a paintenance - a pain to maintain. But I can also see the argument being that git commit args wouldn't change that often.

Reply:
I don't love this either, but I spent a while thinking about it/trying different possible approaches and couldn't come up with anything better.  I think this actually shouldn't introduce an ongoing maintenance burden, for a few reasons:
1) As you mentioned, git commit is pretty stable, so I wouldn't expect flags to be added frequently. (where "not frequently" = "idk maybe on the order of 3 years")
2) This fails safe, so it's okay if the concurrent editor magic doesn't fire when `git commit` is called with a new flag. In terms of user experience interruption, I'd expect new flags to likely be obscure and used pretty rarely.